### PR TITLE
Teach DLM to recognise symbols in COFF Object Files.

### PR DESCRIPTION
**Pull request by gargvaibhav64**
On Windows, the .dll files contain symbol information in COFF Format which DLM currently fails to understand properly. This leads to unrecognized symbol errors while starting ROOT with modules.

This commit teaches DLM to correctly read symbols when the Object File is in COFF executable format.

@vgvassilev @bellenot

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp [hunk 1]
```diff
@@ -541,6 +542,27 @@ namespace cling {
       }
     }
 
+    // On Windows, the symbols are present in COFF format.
+    if (BinObjFile->isCOFF()) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:546]: That should be `else if` to `if (BinObjFile->isELF())`

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp [hunk 2]
```diff
@@ -541,6 +542,27 @@ namespace cling {
...
+      llvm::object::COFFObjectFile* CoffObj = cast<llvm::object::COFFObjectFile>(BinObjFile);
+
+      // In COFF, the symbols are not present in the SymbolTable section
+      // of the Object file. They are present in the ExportDirectory section.
+      export_directory_iterator I = CoffObj->export_directory_begin();
+      export_directory_iterator E = CoffObj->export_directory_end();
+
+      for (; I != E; I = ++I) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:554]: That should become:

`for (auto I = CoffObj->export_directory_begin(), E = CoffObj->export_directory_end();
         I != E; I = ++I) {`

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp [hunk 3]
```diff
@@ -541,6 +542,25 @@ namespace cling {
       }
     }
 
+    // On Windows, the symbols are present in COFF format.
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:545]: Can you move the comment in the condition and stick the `else if` closer?

In addition we should have the same logic for ignoring symbols as we have for ELF -- we need essentially the same code only the loop is over export_directory_begin/end.

Comment (gargvaibhav64) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:545]: I think the inner code is similar? (the logic of ignoring symbols)

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:545]: I fail to see something like:

```
  uint32_t Flags = S.getFlags();
  // DO NOT insert to table if symbol was undefined
  if (Flags & llvm::object::SymbolRef::SF_Undefined)
      continue;
```

in the newly added code...

Comment (gargvaibhav64) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp]: Addressed in a comment...

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp [hunk 4]
```diff
@@ -540,6 +541,25 @@ namespace cling {
...
     }
+    else if (BinObjFile->isCOFF()) { // On Windows, the symbols are present in COFF format.
+      llvm::object::COFFObjectFile* CoffObj = cast<llvm::object::COFFObjectFile>(BinObjFile);
+
+      // In COFF, the symbols are not present in the SymbolTable section
+      // of the Object file. They are present in the ExportDirectory section.
+      for (auto I=CoffObj->export_directory_begin(), 
+                E=CoffObj->export_directory_end(); I != E; I = ++I) {
+        // All the symbols are already flagged as exported. 
+        // Therefore, no check if symbol is uninitialized.
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:552]:
```suggestion
        // We cannot really ignore symbols based on flags as we do on unix.
```

**Pull request merged**